### PR TITLE
[TF 2.0 API Docs] Docstring for tf.train.experimental.enable_mixed_precision_graph_rewrite

### PR DESCRIPTION
**Pull request by tlkh**
In response to #29241

Improved the docstring for `tf.train.experimental.enable_mixed_precision_graph_rewrite`:

* Added example for using function
* Added colab notebook to demonstrate speed-up without performance penalty
* Added original graphic for loss scaling ([source](https://docs.google.com/presentation/d/1ZCREpnt2H7I7J6Xek-Z0bZ_qE3bIHFzbTUed8Vdw6Ag/edit?usp=sharing))
* Added more information about graph rewrite operation
* Added performance guide
* Added exception information
* Added more clarification to `loss_scale` argument

A gist with the rendered docstring is [here](https://gist.github.com/tlkh/fa20c5bf3c8b48def4501cccff8b3559) for ease of review.

Thank you, any feedback or criticism is welcome.

### CODE REVIEW

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 1]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
-  is required to prevent underflow in the float16 tensors during the backwards
-  pass. An optimizer must be passed to this function, which will then be wrapped
-  to use loss scaling.
+  """Enables faster training on compatible hardware accelerators by using a
+  mix of numerical data types. Known as mixed precision, this is achieved
+  via a graph rewrite operation and a scaled optimizer.
+  
+  Mixed precision is the use of both float16 and float32 tensors when training
+  a model. Performing only certain arithmetic operations in lower precision
+  (float16) results in performance gains of when using compatible hardware
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: > performance gains of when

I think the "of" is a mistake

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 2]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
-  This function additionally wraps an Optimizer with a LossScaleOptimizer, which
-  is required to prevent underflow in the float16 tensors during the backwards
-  pass. An optimizer must be passed to this function, which will then be wrapped
-  to use loss scaling.
+  """Enables faster training on compatible hardware accelerators by using a
+  mix of numerical data types. Known as mixed precision, this is achieved
+  via a graph rewrite operation and a scaled optimizer.
+  
+  Mixed precision is the use of both float16 and float32 tensors when training
+  a model. Performing only certain arithmetic operations in lower precision
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: Maybe add a tiny bit explaining why only certain and not all (eg some operations require higher precision)

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 3]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
+  [Colab notebook](https://colab.research.google.com/drive/1qabCQKJZhdIDOrXuhoLa9uMB4RJJ2a1N).
+
+  Calling `enable_mixed_precision_graph_rewrite(opt)` enables the graph rewrite
+  operation before computing gradients. The function additionally returns an
+  `Optimizer`(`opt`) wrapped with a `LossScaleOptimizer`. This prevents underflow
+  in the float16 tensors during the backward pass. An optimizer of type
+  `tf.train.Optimizer` or `tf.keras.optimizers.Optimizer` must be passed to this
+  function, which will then be wrapped to use loss scaling.
+
+  <img src="https://i.imgur.com/5kA9pcS.png" width="500px">
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: I think someone will have to re-host this under tf.org

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 4]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
+                metrics=["accuracy"])
+
+  model.fit(x_train, y_train,
+            batch_size=batch_size,
+            epochs=epochs)
+  ```
+
+  For a complete example showing the speed-up on training an image classification
+  task on CIFAR10, check out this
+  [Colab notebook](https://colab.research.google.com/drive/1qabCQKJZhdIDOrXuhoLa9uMB4RJJ2a1N).
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: The notebook is here: https://github.com/tensorflow/examples/pull/65
The link should probably be updated to the tensorflow/examples repo once that PR is merged

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 5]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
+
+  <img src="https://i.imgur.com/5kA9pcS.png" width="500px">
+
+  The graph rewrite operation changes the `dtype` of certain operations in the
+  graph from float32 to float16. There are several categories of operations that
+  are considered or rejected by this rewrite operation, and they are defined in
+  [auto_mixed_precision_lists.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h),
+  and each category is summarized below:
+
+  * [`ClearList`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h#L178): Ops that do not have numerically significant effects. E.g. `ArgMax` and `Floor`.
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: The line numbers of these links will change over time as new ops are added. Maybe the text should explain that each corresponds to the function in that file.

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 6]
```diff
@@ -73,39 +73,96 @@ def _wrap_optimizer(opt, loss_scale):
...
   enabled within `tf.function`s, as outside `tf.function`s, there is no graph.
 
-  When enabled, mixed precision is only used on Volta GPUs and above. The parts
+  **Performance Guide**
+
+  * Dimensions should be in multiples of 8 where possible.
+  * Dimensions should at least 64, and where smaller than 256, be a power of 2.
+  * Operations with dimensions less than 128 are memory-bound rather than compute-bound.
+  * Batch size should be a multiple of SM count (e.g. a multiple of 80 on V100 GPUs).
+  * Where needed, padding the tensor to a multiple of 8 will result in better performance.
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: put this line second (or third), so it comes together with the other bullets about tensor dimensions. right now its after batch size so is a little confusing.

**Review by perfinion — CHANGES_REQUESTED** [review]
Great work! Just a few small things to tweak :)

**Comment by perfinion** [comment]
Can you change the colours in the image as well? On my monitor the green and blue look fairly similar its hard to distinguish.

**Comment by tlkh** [comment]
I have made all the requested changes!

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 7]
```diff
@@ -73,39 +73,100 @@ def _wrap_optimizer(opt, loss_scale):
...
+  [Colab notebook](https://colab.research.google.com/drive/1qabCQKJZhdIDOrXuhoLa9uMB4RJJ2a1N).
+
+  Calling `enable_mixed_precision_graph_rewrite(opt)` enables the graph rewrite
+  operation before computing gradients. The function additionally returns an
+  `Optimizer`(`opt`) wrapped with a `LossScaleOptimizer`. This prevents underflow
+  in the float16 tensors during the backward pass. An optimizer of type
+  `tf.train.Optimizer` or `tf.keras.optimizers.Optimizer` must be passed to this
+  function, which will then be wrapped to use loss scaling.
+
+  <img src="https://i.imgur.com/LYVEbcN.png" width="500px">
```
Comment (perfinion) [inline tensorflow/python/training/experimental/mixed_precision.py]: Note to whoever merges this that this image should be re-hosted somewhere

**Review by perfinion — DISMISSED** [review]

**Comment by perfinion** [comment]
@rthadur Can you make sure the links are updated when this is merged?

**Comment by tlkh** [comment]
Fixed the pylint 80 char line limit errors.

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 8]
```diff
@@ -73,39 +73,116 @@ def _wrap_optimizer(opt, loss_scale):
...
-  precision to speed up the execution time of your model when run on a GPU. It
-  does this by changing the dtype of certain operations in the graph from
-  float32 to float16.
-
-  This function additionally wraps an Optimizer with a LossScaleOptimizer, which
-  is required to prevent underflow in the float16 tensors during the backwards
-  pass. An optimizer must be passed to this function, which will then be wrapped
-  to use loss scaling.
+  """Enables faster training on compatible hardware accelerators by using a
+  mix of numerical data types. Known as mixed precision, this is achieved
```
Comment (martinwicke) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please make sure the brief description is only a single line.

Comment (tlkh) [reply tensorflow/python/training/experimental/mixed_precision.py]: Just to clarify, the brief description has to be within 80 characters?

Comment (martinwicke) [reply tensorflow/python/training/experimental/mixed_precision.py]: Actually a bit less because of the indent and """. But yes, the first line of the docstring is brief, then a blank line, then the description.

Comment (tlkh) [reply tensorflow/python/training/experimental/mixed_precision.py]: I have made the requested changes!

**Comment by tlkh** [comment]
Pushed the amended commit with the requested changes.

**Comment by recrusader** [comment]
@tlkh, when enabling "enable_mixed_precision_graph_rewrite", how about the data type of model network when defining the model?
still use FP32? if it is, how to output the FP16 trained model? Thanks!

**Comment by tlkh** [comment]
> @tlkh, when enabling "enable_mixed_precision_graph_rewrite", how about the data type of model network when defining the model?
> still use FP32? if it is, how to output the FP16 trained model? Thanks!

You define your model as per normal in FP32. Before training (computing of gradients), your graph will be "rewritten" to have some nodes in FP16, others in FP32. This is **only used for training**, and if you save or export your model, it will be in FP32 as per normal.

This means that training is faster, but otherwise no behavior differs. You will not be able to export the actual "mixed precision model", but you don't have to since it doesn't run faster on CPU, and you can also run the graph rewrite before the next training session.

**Comment by recrusader** [comment]
Thank you very much @tlkh. However, FP16 model can run faster compared to the FP32 one on GPU. you mentioned "you can also run the graph rewrite before the next training session." With this method, FP16 model can be saved? How to do this?

Is there no way to save the FP16 model with auto-mixed precision training? thanks a lot.

**Comment by tlkh** [comment]
> Thank you very much tlkh. However, FP16 model can run faster compared to the FP32 one on GPU. you mentioned "you can also run the graph rewrite before the next training session." With this method, FP16 model can be saved? How to do this?
> Is there no way to save the FP16 model with auto-mixed precision training? thanks a lot.

There is no way to save the FP16 model, but **there is no need to**!

Export the model as per normal. The next time you want to use it, just run the graph rewrite again. If loading it for inference, you can manually set it in the TensorFlow session:

```python
# TF1.x
config = tf.ConfigProto()
config.graph_options.rewrite_options.auto_mixed_precision = True
sess = tf.Session(config=config)

# TF2.x
tf.config.optimizer.set_experimental_options({
    "auto_mixed_precision": True
})
```

**Comment by rthadur** [comment]
@perfinion @martinwicke can you please review new changes.

**Comment by recrusader** [comment]
Thanks a lot, @tlkh. However, if FP16 trained network cannot be saved, it is difficult to use the network outside tensorflow after finishing the training.

**Comment by recrusader** [comment]
In addition, it seems that ops conversion from FP32 to FP16 can happen on new GPU, like Volta. TF cannot do this conversion for old one.

**Comment by tlkh** [comment]
> Thanks a lot, tlkh. However, if FP16 trained network cannot be saved, it is difficult to use the network outside tensorflow after finishing the training.
> In addition, it seems that ops conversion from FP32 to FP16 can happen on new GPU, like Volta. TF cannot do this conversion for old one.

If you wish to use the network outside of TensorFlow for inference in FP16, typically just converting all the datatypes to FP16 will work. Only for training does there have to be special considerations. There are other toolkits to help with inference optimization outside of TensorFlow, such as [TensorRT](https://developer.nvidia.com/tensorrt).

Only Volta and Turing GPUs have the FP16 units and Tensor Cores that can benefit from mixed precision. Older GPUs will not see a benefit, hence the feature is not enabled for them.

@recrusader let's move this discussion to a new GitHub issue if you wish to raise any concerns or suggestions. Keep this thread for discussion specifically about this PR.

**Comment by recrusader** [comment]
I asked this question in official model. However, no one can give me an answer. Thank you very much! I think that I have got the answer.

**Comment by tlkh** [comment]
@perfinion @martinwicke requesting review once again.

Changes made:

* image and example notebook is now hosted by NVIDIA
* docstring wording has been improved, and updated/merged with recent changes from master

Thank you!

**Review by martinwicke — COMMENTED** [review]
@MarkDaoust will this render correctly? I see there's no indent on the args and returns sections, I'm wondering whether the indent is required.

**Comment by tlkh** [comment]
@martinwicke thanks for pointing that out. Let me fix that anyway.

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 9]
```diff
@@ -77,43 +77,125 @@ def _wrap_optimizer(opt, loss_scale, use_v1_behavior):
...
-
-  WARNING: This rewrite silently affects the entire model and can have
-  unintended consequences. One example: If a NaN occurs during dynamic loss
-  scaling, the data for the batch is silently dropped while the
-  LossScaleOptimizer attempts to find the appropriate scaling value on the next
-  batch.
-
+  
+  Raises:
+  `ValueError` when
```
Comment (MarkDaoust) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please Indent

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 10]
```diff
@@ -77,43 +77,125 @@ def _wrap_optimizer(opt, loss_scale, use_v1_behavior):
...
+  `tf.train.experimental.enable_mixed_precision_graph_rewrite()`
+  is called.
+  
   Args:
-    opt: An instance of a `tf.keras.optimizers.Optimizer`.
-    loss_scale: Either an int/float, the string "dynamic", or an instance of a
-      `tf.train.experimental.LossScale`. The loss scale to use. It is
-      recommended to keep this as its default value of "dynamic".
-
+  opt: An instance of a `tf.keras.optimizers.Optimizer`.
```
Comment (MarkDaoust) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please indent

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 11]
```diff
@@ -123,38 +205,124 @@ def enable_mixed_precision_graph_rewrite(opt, loss_scale='dynamic'):
...
+  
+  Currently, mixed precision is only enabled on NVIDIA Tensor Core GPUs with
+  Compute Capability 7.0 and above (Volta, Turing, or newer architectures). The
+  parts of the graph on CPUs and TPUs are untouched by the graph rewrite. TPU
   support is coming soon. CPUs are not supported, as CPUs do not run float16
   operations faster than float32 operations.
-
+  
+  Raises:
+  `ValueError` when
```
Comment (MarkDaoust) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please indent

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 12]
```diff
@@ -123,38 +205,124 @@ def enable_mixed_precision_graph_rewrite(opt, loss_scale='dynamic'):
...
+  is called.
+  
   Args:
-    opt: An instance of a `tf.keras.optimizers.Optimizer` or a
-      `tf.train.Optimizer`.
-    loss_scale: Either an int/float, the string "dynamic", or an instance of a
-      `tf.train.experimental.LossScale`. The loss scale to use. It is
-      recommended to keep this as its default value of "dynamic".
-
+  opt: An instance of a `tf.keras.optimizers.Optimizer` or a
```
Comment (MarkDaoust) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please Indent

File: tensorflow/python/training/experimental/mixed_precision.py [hunk 13]
```diff
@@ -123,38 +205,124 @@ def enable_mixed_precision_graph_rewrite(opt, loss_scale='dynamic'):
...
+    `tf.train.Optimizer`.
+  loss_scale: Either an int/float, the string `"dynamic"`, or an instance of
+    a `tf.train.experimental.LossScale`. The loss scale to use. It is
+    recommended to keep this as its default value of `"dynamic"`, which will 
+    adjust the scaling automatically to prevent `Inf` or `NaN` values.
+  
   Returns:
-    A version of `opt` that will use loss scaling to prevent underflow.
+  A version of `opt` that will use loss scaling to prevent underflow.
```
Comment (MarkDaoust) [inline tensorflow/python/training/experimental/mixed_precision.py]: Please indent.

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
Hi @tlkh, 

It looks like you miss understood @martinwicke's comment.

All the Args/Returns/Raises blocks need to be indented, or they will not be recognized by our linters, and will render incorrectly on tensorflow.org (They will be interpreted as markdown which will just flatten them into a single paragraph).

Thanks.

**Comment by tlkh** [comment]
Sorry about that! I fixed it.

**Review by martinwicke — APPROVED** [review]

**Comment by tlkh** [comment]
> Hi @tlkh,
> 
> It looks like you miss understood @martinwicke's comment.
> 
> All the Args/Returns/Raises blocks need to be indented, or they will not be recognized by our linters, and will render incorrectly on tensorflow.org (They will be interpreted as markdown which will just flatten them into a single paragraph).
> 
> Thanks.

Hmm I've addressed the changes but I'm not sure why GitHub is still complaining about "1 change requested".
